addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiSocket.java

```diff
@@ -0,0 +1,203 @@
+     */
+    public synchronized void registerListener(XiaomiSocketListener listener) {
+        if (socket == null) {
+            setupSocket();
+        }
+        if (!listeners.contains(listener)) {
+            logger.trace("Adding socket listener {}", listener.toString());
+            listeners.add(listener);
+        }
+    }
+
+    /**
+     * Unregisters a {@link XiaomiSocketListener}. If there are no listeners left,
+     * the {@link XiaomiSocket} is being closed.
+     *
+     * @param listener - {@link XiaomiSocketListener} to be unregistered
+     */
+    public synchronized void unregisterListener(XiaomiSocketListener listener) {
+        listeners.remove(listener);
+
+        if (listeners.isEmpty()) {
+            closeSocket();
+        }
+    }
+
+    /**
+     * Sends a message through the {@link XiaomiSocket} to a specific address and port
+     *
+     * @param message - Message to be sent
+     * @param address - Address, to which the message shall be sent
+     * @param port - - Port, through which the message shall be sent
+     */
+    public void sendMessage(String message, InetAddress address, int port) {
+        try {
+            byte[] sendData = message.getBytes("UTF-8");
+            DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, address, port);
+            logger.trace("Sending message: {} to {}", message, address.toString());
+            socket.send(sendPacket);
+        } catch (IOException e) {
+            logger.error("Sending error", e);
+        }
+    }
+
+    /**
+     * @return - the port number of this {@link XiaomiSocket}
+     */
+    public int getPort() {
+        return port;
+    }
+
+    /**
+     * @return - a list of already open sockets
+     */
+    public static ArrayList<XiaomiSocket> getOpenSockets() {
+        return openSockets;
+    }
+
+    /**
+     * The thread, which waits for data on the {@link XiaomiSocket} and handles it, when received
+     *
+     * @author Patrick Boos - Initial contribution
+     * @author Dimalo - comments and synchronized block for callback instead of copy
+     *
+     */
+    class ReceiverThread extends Thread {
+        @Override
+        public void run() {
+            logger.trace("Staring reveicer thread for socket on port {}", socket.getLocalPort());
+            receiveData(socket, datagramPacket);
+        }
+
+        /**
+         * This method is the main method of the {@link ReceiverThread} for the {@link XiaomiBridgeSocket}.
+         * If the socket has data, it parses the data to a json object and calls all
+         * {@link XiaomiSocketListener} and passes the data to them.
+         *
+         * @param socket - The multicast socket to listen to
+         * @param dgram - The datagram to receive
+         */
+        private void receiveData(DatagramSocket socket, DatagramPacket dgram) {
+            try {
+                while (true) {
```

Comment:
So your statement is that if the Thread is interrupted it will start throwing IOExceptions? I don't know if that is the case, I do indeed see that you check it when you have an IOException but I doubt that you will get the IOException.